File: leetcode/tree/minimum_depth_binary_tree.go
package main

import "fmt"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func minDepth(root *TreeNode) int {
	if root == nil {
		return 0
	}
	out := 999999999
	solve(root, 1, &out)
	return out
}

func solve(root *TreeNode, level int, out *int) {
	if root == nil {
		return
	}

	if root.Left == nil && root.Right == nil {
		if *out > level {
			*out = level
		}
		return
	}
	solve(root.Left, level+1, out)
	solve(root.Right, level+1, out)
}

func main() {
	root := &TreeNode{1, nil, nil}
	root.Left = &TreeNode{2, nil, nil}
	root.Right = &TreeNode{2, nil, nil}
	root.Left.Left = &TreeNode{3, nil, nil}
	// root.Left.Right = &TreeNode{4, nil, nil}
	root.Right.Left = &TreeNode{4, nil, nil}
	root.Right.Right = &TreeNode{3, nil, nil}
	min := minDepth(root)
	fmt.Println(min)
}

File: leetcode/tree/minimum_depth_binary_tree.py
from common import TreeNode

class Solution():
    def __init__(self):
        self.out = 999999999
    def minDepth(self, root: TreeNode) -> int:
        def solve(root: TreeNode, l):
            if not root:
                return
            if not root.left and not root.right:
                self.out = min(self.out, l)
                return
            solve(root.left, l + 1)
            solve(root.right, l + 1)
        if not root:
            return 0
        solve(root, 1)
        return self.out
    
if __name__ == "__main__":
    root = TreeNode(1)
    root.left = TreeNode(2)
    root.right = TreeNode(3)
    root.right.left = TreeNode(4)
    root.right.right = TreeNode(5)
    print(Solution().minDepth(root))